File: src/v1/languages/javascript/templates/universal_wrapper.js
// === Sandbox-Link JavaScript Wrapper for QuickJS ===
import * as std from 'std';

const START_MARKER = "__SANDBOX_RESULT_START__";
const END_MARKER = "__SANDBOX_RESULT_END__";

function sendResult(msg) {
  const out = JSON.stringify(msg);
  // Default to stdout with markers
  std.out.puts(START_MARKER + out + END_MARKER + "\n");
  std.out.flush();
}

// Simple Log Interceptor for QuickJS
function setupLogInterception() {
  const originalLog = console.log;
  console.log = function(...args) {
    const content = args.map(arg => typeof arg === 'object' ? JSON.stringify(arg) : String(arg)).join(' ');
    sendResult({
      ver: '1.0',
      id: 'log-evt',
      type: 'log',
      stream: 'stdout',
      content: content + "\n"
    });
    originalLog.apply(console, args);
  };
}

async function main() {
  setupLogInterception();
  const stdin = std.stdin || std.in || std.fdopen(0, "r");
  if (!stdin) return;

  const mByte = stdin.getByte();
  if (mByte < 0) return;
  const mode = String.fromCharCode(mByte);
  
  let lenStr = '';
  for (let i = 0; i < 8; i++) {
    const b = stdin.getByte();
    if (b < 0) break;
    lenStr += String.fromCharCode(b);
  }
  const length = parseInt(lenStr, 16);

  if (mode === 'A') {
    let payload = '';
    for (let i = 0; i < length; i++) {
      const byte = stdin.getByte();
      if (byte < 0) break;
      payload += String.fromCharCode(byte);
    }

    const callMsg = JSON.parse(payload);
    const { id, method, params } = callMsg;
    
    let result;
    try {
      const userModule = await import('./user_code.js');
      const func = userModule[method];
      if (typeof func !== 'function') {
        throw new Error(`Function '${method}' not found`);
      }

      const { args = [], kwargs = {} } = params || {};
      if (Object.keys(kwargs).length > 0) {
        result = await func(...args, kwargs);
      } else {
        result = await func(...args);
      }

      sendResult({
        ver: '1.0',
        id: id,
        type: 'result',
        status: 'ok',
        data: { result }
      });
    } catch (e) {
      sendResult({
        ver: '1.0',
        id: id,
        type: 'result',
        status: 'fail',
        data: {
          error: {
            message: e.message || String(e),
            type: e.name || 'Error',
            stack: e.stack
          }
        }
      });
    }
  }
}

main().catch(err => {
  sendResult({
    ver: '1.0',
    type: 'result',
    status: 'fail',
    data: {
      error: { message: "Fatal: " + (err.message || String(err)), type: 'FatalError' }
    }
  });
});

File: src/v1/languages/c/templates/universal_wrapper.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "cJSON.h"

#define START_MARKER "__SANDBOX_RESULT_START__"
#define END_MARKER "__SANDBOX_RESULT_END__"

cJSON* __sandbox_dispatch(cJSON* params);

void send_result(cJSON *result_msg) {
    char *out_str = cJSON_PrintUnformatted(result_msg);
    const char *result_fd_env = getenv("SB_RESULT_FD");
    
    if (result_fd_env != NULL) {
        int fd = atoi(result_fd_env);
        // In some WASM environments, fdopen might not be available or behaves differently
        // but we keep it for native/future compatibility
        FILE *fp = fdopen(fd, "w");
        if (fp != NULL) {
            fprintf(fp, "%s\n", out_str);
            fclose(fp);
            free(out_str);
            return;
        }
    }
    
    // Default to stdout with markers
    printf("%s%s%s\n", START_MARKER, out_str, END_MARKER);
    free(out_str);
}

int main() {
    char header[10];
    for (int i = 0; i < 9; i++) {
        int c = getchar();
        if (c == EOF) return 0;
        header[i] = (char)c;
    }
    header[9] = '\0';

    char mode = header[0];
    unsigned int length = (unsigned int)strtoul(&header[1], NULL, 16);

    if (mode == 'A') {
        char *buffer = (char *)malloc(length + 1);
        for (unsigned int i = 0; i < length; i++) {
            int c = getchar();
            if (c == EOF) break;
            buffer[i] = (char)c;
        }
        buffer[length] = '\0';

        cJSON *call_msg = cJSON_Parse(buffer);
        if (call_msg == NULL) {
            cJSON *res = cJSON_CreateObject();
            cJSON_AddStringToObject(res, "ver", "1.0");
            cJSON_AddStringToObject(res, "type", "result");
            cJSON_AddStringToObject(res, "status", "fail");
            cJSON *data = cJSON_AddObjectToObject(res, "data");
            cJSON *error = cJSON_AddObjectToObject(data, "error");
            cJSON_AddStringToObject(error, "message", "Failed to parse call message");
            cJSON_AddStringToObject(error, "type", "ParseError");
            send_result(res);
            cJSON_Delete(res);
            free(buffer);
            return 0;
        }

        // Get call ID to reply with same ID
        cJSON *id_node = cJSON_GetObjectItem(call_msg, "id");
        const char *id_str = (id_node != NULL) ? id_node->valuestring : "unknown";

        cJSON *params_node = cJSON_GetObjectItem(call_msg, "params");

        cJSON *execution_data = __sandbox_dispatch(params_node);
        
        cJSON *res = cJSON_CreateObject();
        cJSON_AddStringToObject(res, "ver", "1.0");
        cJSON_AddStringToObject(res, "id", id_str);
        cJSON_AddStringToObject(res, "type", "result");
        cJSON_AddStringToObject(res, "status", "ok");
        cJSON_AddItemToObject(res, "data", execution_data);
        
        send_result(res);

        free(buffer);
        cJSON_Delete(call_msg);
        cJSON_Delete(res);
    }
    return 0;
}

#include "cJSON.c"
#include "__sandbox_dispatcher.c"
